python/tank_vendor/shotgun_authentication/interactive_authentication.py

```diff
@@ -0,0 +1,432 @@
+# Copyright (c) 2015 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+UI and console based login for Toolkit.
+"""
+
+# Using "with" with the lock to make sure it is always released.
+
+from __future__ import with_statement
+from getpass import getpass
+import threading
+from .errors import AuthenticationError, AuthenticationDisabled, AuthenticationCancelled
+from . import session_cache
+
+
+# FIXME: Quick hack to easily disable logging in this module while keeping the
+# code compatible. We have to disable it by default because Maya will print all out
+# debug strings.
+if False:
+    # Configure logging
+    import logging
+    logger = logging.getLogger("sgtk.interactive_authentication")
+    logger.setLevel(logging.DEBUG)
+    logger.addHandler(logging.StreamHandler())
+else:
+    class logger:
+        @staticmethod
+        def debug(*args, **kwargs):
+            pass
+
+        @staticmethod
+        def info(*args, **kwargs):
+            pass
+
+        @staticmethod
+        def warning(*args, **kwargs):
+            pass
+
+        @staticmethod
+        def error(*args, **kwargs):
+            pass
+
+        @staticmethod
+        def exception(*args, **kwargs):
+            pass
+
+
+def _get_qt_state():
+    """
+    Returns the state of Qt: the librairies available and if we have a ui or not.
+    :returns: If Qt is available, a tuple of (QtCore, QtGui, has_ui_boolean_flag).
+              Otherwise, (None, None, None)
+    """
+    try:
+        from .ui.qt_abstraction import QtGui, QtCore
+    except ImportError:
+        return None, None, None
+    return QtCore, QtGui, QtGui.QApplication.instance() is not None
```

Comment:
This looks fine but I would proceed with more caution. Over the years we have seen lots of strange reasons why QT fails to import (not always import errors!). Calling `QtGui.QApplication.instance()` may also fail because there are dso/dll problems for example. We don't want this method to ever raise exceptions, right? I suggest a more aggressive pokemon style approach here. Maybe something like

```
   qt_core = None
   qt_gui = None
   qt_instance_active = False
   try:
        from .ui.qt_abstraction import QtGui, QtCore
        qt_instance_active = (QtGui.QApplication.instance() is not None)
    except:
        pass
    return (qt_core, qt_gui, qt_instance_active)
```